File: Python/houseRobber.py
class Solution:
    def rob(self, nums: list) -> int:
        """
        O(n)
        :param nums:
        :return:
        """
        dp = nums + [0,0,0]
        for i in range(len(nums) - 1, -1,-1):
            dp[i] = dp[i] + max(dp[i + 2], dp[i + 3])
        return max(dp)

    def rob2(self, nums: list ) -> int:
        """
        O(n)

        :param nums:
        :return:
        """
        l1_max = self.rob(nums[1:])
        l2_max = self.rob(nums[:-1])
        return max(l1_max, l2_max)


if __name__ == "__main__":
    nums = [1,2,3,1]
    ans = 4
    resp = Solution().rob2(nums)
    if ans == resp:
        print("PASS", resp)
    else:
        print("FAIL", resp)

File: Java/src/HouseRobber.java
class Solution {
    public int rob(int[] nums) {
        if (nums.length == 2){
            return Math.max(nums[0], nums[1]);
        }
        if (nums.length  == 1){
            return nums[0];
        }
        if (nums.length == 0 ){
            return -1;
        }
        int[] dp = new int[nums.length + 1]; 
        Arrays.fill(dp, 0); 
        
        for (int i = 0 ; i < nums.length; i++){
            int b1 = Math.max(0, i - 1); 
            int b2 = Math.max(0, i - 2);
            
            dp[i + 1] = Math.max(
                nums[i] + dp[b1], 
                nums[i] + dp[b2]
            );
        }
        return Math.max(dp[dp.length - 1 ], dp[dp.length - 2]); 
        
        
    }
}
